mox/impl/src/lib.rs

```diff
@@ -1,245 +1,239 @@
+                Err(Error::new(punctuated.span(), "Colon tag name syntax isn't supported"))
+            }
+        }
     }
 }
 
-impl ToTokens for MoxTag {
-    fn to_tokens(&self, tokens: &mut TokenStream) {
-        tag_to_tokens(&self.name, &self.attributes, Some(&self.children), tokens);
+fn mangle_expr_path(name: &mut syn::ExprPath) {
+    for segment in name.path.segments.iter_mut() {
+        mangle_ident(&mut segment.ident);
     }
 }
 
-fn tag_to_tokens(
-    name: &Ident,
-    attributes: &[MoxAttr],
-    children: Option<&[MoxItem]>,
-    stream: &mut TokenStream,
-) {
-    // this needs to be nested within other token groups, must be accumulated
-    // separately from stream
-    let mut contents = quote!();
+impl TryFrom<syn_rsx::Node> for MoxAttr {
+    type Error = syn::parse::Error;
 
-    for attr in attributes {
-        attr.to_tokens(&mut contents);
-    }
-
-    if let Some(items) = children {
-        for child in items {
-            MoxAttr::child(child).to_tokens(&mut contents);
+    fn try_from(node: syn_rsx::Node) -> syn::Result<Self> {
+        use syn_rsx::NodeType::*;
+        match node.node_type {
+            Element | Text | Block => Err(node_parse_error(&node, "MoxAttr")),
+            Attribute => {
+                Ok(MoxAttr { name: MoxAttr::validate_name(node.name.unwrap())?, value: node.value })
+            }
         }
     }
-
-    quote!(mox::topo::call(|| { #name() #contents .build() })).to_tokens(stream);
 }
 
-impl From<SnaxTag> for MoxTag {
-    fn from(SnaxTag { name, attributes, children }: SnaxTag) -> Self {
-        let attributes = attributes.into_iter().map(MoxAttr::from).collect();
-        Self {
-            // TODO get the span for the whole tag
-            span: name.span(),
-            name,
-            attributes,
-            children: children.into_iter().map(MoxItem::from).collect(),
+impl MoxAttr {
+    fn validate_name(name: syn_rsx::NodeName) -> syn::Result<syn::Ident> {
+        use syn::{parse::Error, punctuated::Pair, PathSegment};
+        use syn_rsx::NodeName;
+
+        let invalid_error = |span| Error::new(span, "Invalid name for an attribute");
+
+        match name {
+            NodeName::Path(syn::ExprPath {
+                attrs,
+                qself: None,
+                path: syn::Path { leading_colon: None, mut segments },
+            }) if attrs.is_empty() && segments.len() == 1 => {
+                let pair = segments.pop();
+                match pair {
+                    Some(Pair::End(PathSegment { mut ident, arguments }))
+                        if arguments.is_empty() =>
+                    {
+                        mangle_ident(&mut ident);
+                        Ok(ident)
+                    }
+                    _ => Err(invalid_error(segments.span())),
+                }
+            }
+            NodeName::Dash(punctuated) => {
+                Err(Error::new(punctuated.span(), "Dash attribute name syntax isn't supported"))
```

Comment:
If it's possible to support with syn-rsx, that's another reason in favor of making this change. There are native HTML attributes like `aria-valuemin` that would be nice to spell identically as on the web, and there's also the web's use of `data-*` attributes which we have a hard time supporting right now.

It would be OK to do as a follow-up to this change, if you want to do it that way could you file an issue and leave a TODO?